compiler/optimizer/VPHandlers.cpp

```diff
@@ -2019,77 +2028,114 @@ static TR::Node *findArrayIndexNode(OMR::ValuePropagation *vp, TR::Node *node, i
   bool usingAladd = (vp->comp()->target().is64Bit()
                      ) ?
           true : false;
+   // Note: index node here refers to the array element index.
+   if (node->getFirstChild()->isDataAddrPointer())
+      {
+      TR::Node *offset2 = NULL;
+      if (offset->getOpCode().isShiftLogical() ||
+         offset->getOpCode().isMul())
+         {
+         offset2 = offset;
+         }
+      else
+         offset2 = offset->getFirstChild();
```

Comment:
More comments about the shape of the trees in these two cases, the 'if' and 'else' here would be good. For example, why is it okay to proceed with `offset2 == offset` in the `if` case ? More importantly, why is it okay to call `getFirstChild` on `offset` here ? i.e. do we know that `offset` has children ?

Reply:
Here's some context for that logic:
Expected tree structure
1. Some constant needs to be added to the offset
```
aladd
    aloadi (dataAddrPointer)
        aload (array object)
    ladd/lsub
        lmul
            i2l
                iload array_index
            lconst mul_stride/shift_amount
    lconst some_constant
```
2. No constant needs to be added to the offset
```
aladd
    aloadi (dataAddrPointer)
        aload (array object)
    lmul
        i2l
            iload array_index
        lconst mul_stride
```
3. Stride == 1
```
aladd
    aloadi (dataAddrPointer)
        aload (array object)
    iload array_index
```
`Offset2` is just an alias for `offset` node which points to `lmul/lshl` or `array_index` node.

We can proceed with `offset2 == offset` because we assume that either it is array_index (case 3) or it has array_index as a child of `lmul` node.

> why is it okay to call getFirstChild on offset on the else side ? i.e. do we know that offset has children ?

We expect offset to have a child based on the expected tree structures. Either the child is `lmul` or `array_index` node.

We should add a check for these assumptions. @rmnattas While we are at it can you add a assert failure for all other node types? I added a check for shift node but it doesn't look like it is handled correctly here.

Reply:
Thank you @VermaSh, not seeing what case would result in 1. (some constant being added beside offset) as non-offheap case doesn't cover that?

Reply:
Added comments to explain the trees in the code.